### PR TITLE
feat: sign in page markup od-25

### PR DESCRIPTION
**Pull request by kovalsofiia1**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -58,6 +58,7 @@
 		"webpack-cli": "^5.1.4"
 	},
 	"dependencies": {
+		"@types/bcrypt": "^5.0.2",
```
Comment (GEOFARL) [inline package.json:61]: It should be a devDependency

File: src/views/pages/sign-in.ejs [hunk 2]
```diff
@@ -6,30 +6,38 @@
...
+                }); %>
+              </div>
             </form>
-            <a href="/sign-up">Зареєструватися</a>
+           
       </div>
     </div>
 
     <script src="/js/sign-in.js"></script>
+    <script src="/js/toggle-password.js"></script>
```
Comment (GEOFARL) [inline src/views/pages/sign-in.ejs:40]: You should write create this function inside of the `client/src/shared/helpers` folder, use it in the `sign-in.ts` file

Comment (GEOFARL) [reply src/views/pages/sign-in.ejs:40]: js files are produced in the result of the compilation process and they're ignored, so you even has not commited it

File: src/views/pages/sign-in.ejs [hunk 3]
```diff
@@ -6,30 +6,38 @@
...
+              <hr>
+              <div class="form-button-container">
+                <%- include('../components/button.ejs', {
+                  type: "submit",
+                  label: "Увійти"
+                }); %>
+              </div>
             </form>
-            <a href="/sign-up">Зареєструватися</a>
+
```
Comment (GEOFARL) [inline src/views/pages/sign-in.ejs:35]:
```suggestion
```

File: src/views/components/password-input.ejs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<div class="password-input-container">
+    <%- include('../components/input.ejs', {
+      id: "password",
+      label: "Password",
+      type: "password",
+      placeholder: "***********",
+    }); %>
+    <button class="toggle-password-btn" id="togglePassword">
```
Comment (GEOFARL) [inline src/views/components/password-input.ejs:8]: include type="button", now it submits a form

File: src/views/components/password-input.ejs [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+      <img src="/icons/eye-password.svg" alt="" class="show-password-icon">
+    </button>
+  </div>
```
Comment (GEOFARL) [inline src/views/components/password-input.ejs:11]: do not create a separate component for it

File: src/views/pages/sign-in.ejs [hunk 6]
```diff
@@ -6,30 +6,38 @@
 	<body>
 		<div class="wrapper">
       <div class="container">
-        <h1>З поверненням</h1>
+        <div class="sign-in-welcome-container">
+        <h1 class="sign-in-title">З поверненням!</h1>
+        <div class="register-container">
+          <p class="sign-in-text">Немає акаунту?</p>
+          <a href="/sign-up" class="register-text">Зареєструватися</a>
```
Comment (GEOFARL) [inline src/views/pages/sign-in.ejs:13]: create a link component

File: src/views/pages/sign-in.ejs [hunk 7]
```diff
@@ -6,30 +6,38 @@
 	<body>
 		<div class="wrapper">
       <div class="container">
-        <h1>З поверненням</h1>
+        <div class="sign-in-welcome-container">
+        <h1 class="sign-in-title">З поверненням!</h1>
```
Comment (GEOFARL) [inline src/views/pages/sign-in.ejs:10]:
```suggestion
        <div class="welcome-container">
        <h2 class="heading">З поверненням!</h1>
```
remove sign-in part everywhere, as it's already sign in page and we're using specific sign-in styles

File: src/views/pages/sign-in.ejs [hunk 8]
```diff
@@ -6,30 +6,38 @@
...
+        <div class="register-container">
+          <p class="sign-in-text">Немає акаунту?</p>
+          <a href="/sign-up" class="register-text">Зареєструватися</a>
+        </div>
```
Comment (GEOFARL) [inline src/views/pages/sign-in.ejs:14]:
```suggestion
        <div class="prompt">
          <p class="prompt__question">Немає акаунту?</p>
          <a href="/sign-up" class="prompt__link">Зареєструватися</a>
        </div>
```

File: src/views/components/button.ejs [hunk 9]
```diff
@@ -1,4 +1,4 @@
 <button
   type="<%= locals.type ? type : 'button' %>"
   class="button"
-><%= label %></button>
+><%= label %></button>
```
Comment (GEOFARL) [inline src/views/components/button.ejs:4]: unnecessary change

File: src/client/public/css/pages/sign-in.css [hunk 10]
```diff
@@ -9,4 +9,73 @@
...
+	display: flex;
+	flex-direction: column;
+	gap: 2px;
+	margin-bottom: 18px;
+}
+
+.sign-in-title {
+	text-align: center;
+}
+.register-container {
```
Comment (GEOFARL) [inline src/client/public/css/pages/sign-in.css:25]: keep the spacing consistent

**Review by GEOFARL — CHANGES_REQUESTED** [review]

File: src/views/pages/sign-in.ejs (hunk 8 above)
Comment (GEOFARL) [inline src/views/pages/sign-in.ejs:14]:
```suggestion
        <div class="sign-in-welcome-container">
        <h1 class="sign-in-title">З поверненням!</h1>
        <div class="register-container">
          <p class="sign-in-text">Немає акаунту?</p>
          <a href="/sign-up" class="register-text">Зареєструватися</a>
        </div>
```

keep indentation levels

**Comment by GEOFARL** [comment]
Also, add button hover, active and focus states

File: package.json (hunk 1 above)
Comment (GEOFARL) [inline package.json:61]: it is still in dependencies, not resolved

File: src/client/src/pages/sign-in/sign-in.ts [hunk 11]
```diff
@@ -15,6 +15,8 @@ import {
 	validation,
 } from "~/shared/index.js";
 
+import { initPasswordToggle } from "../../../public/shared/helpers/password-toggle";
```
Comment (GEOFARL) [inline src/client/src/pages/sign-in/sign-in.ts:18]:
```suggestion
import { initPasswordToggle } from "~/shared/index.js";
```

Comment (GEOFARL) [reply src/client/src/pages/sign-in/sign-in.ts:18]: image

File: src/client/public/shared/helpers/password-toggle.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
+export function initPasswordToggle() {
```
Comment (GEOFARL) [inline src/client/public/shared/helpers/password-toggle.ts:1]:
```suggestion
const initPasswordToggle = () => {
```

File: src/client/public/shared/helpers/password-toggle.ts [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+		passwordInput.setAttribute("type", type);
+
+		// Change the icon based on the password visibility
+		if (type === "password") {
+			showPasswordIcon.setAttribute("src", "/icons/eye-closed.svg");
+		} else {
+			showPasswordIcon.setAttribute("src", "/icons/eye-visible.svg");
+		}
+	});
+}
```
Comment (GEOFARL) [inline src/client/public/shared/helpers/password-toggle.ts:20]:
```suggestion
}

export { initPasswordToggle };
```

File: src/views/components/link.ejs [hunk 14]
```diff
@@ -0,0 +1 @@
+<a href="<%= href %>" class="<%= linkClass %>"><%= text %></a>
```
Comment (GEOFARL) [inline src/views/components/link.ejs:1]: It would throw errors, if we do not include link class, make it optional

Comment (GEOFARL) [inline src/views/components/link.ejs:1]:
```suggestion
<a href="<%= href %>" class="<%= className %>"><%= content %></a>
```

File: src/views/pages/sign-in.ejs [hunk 15]
```diff
@@ -6,37 +6,58 @@
...
+                <p class="error-message" data-error-password></p>
+              </div>
+              <hr>
+              <div class="form-button-container">
+                <%- include('../components/button.ejs', {
+                  type: "submit",
+                  label: "Увійти"
+                }); %>
+              </div>
+
```
Comment (GEOFARL) [inline src/views/pages/sign-in.ejs:56]:
```suggestion
```

File: package.json (hunk 1 above)
Comment (GEOFARL) [reply package.json:61]: btw, why do you need at all?

File: src/client/public/css/pages/sign-in.css [hunk 16]
```diff
@@ -9,6 +9,72 @@
...
+	position: absolute;
+	top: 65%;
+	transform: translateY(-50%);
+	right: 16px;
+	cursor: pointer;
+}
+
+.show-password-icon {
+	width: 24px;
+	height: 24px;
```
Comment (GEOFARL) [inline src/client/public/css/pages/sign-in.css:77]: image

remove extra spacing between form controls, extra paddings, etc make it look as in design

File: src/client/public/css/components/button.css [hunk 17]
```diff
@@ -9,4 +9,15 @@
 	background-color: var(--color-red);
 	font-size: var(--font-size-700);
 	line-height: 24px;
+	width: 100%;
+	font-weight: 500;
+}
```
Comment (GEOFARL) [inline src/client/public/css/components/button.css:14]: add transition and pointer cursor

File: src/views/pages/sign-in.ejs [hunk 18]
```diff
@@ -6,37 +6,58 @@
...
+                      "data-input-password": ""
+                    }
+                  }); %>
+                  <button class="toggle-password-btn" id="togglePassword" type="button">
+                    <img src="/icons/eye-closed.svg" alt="" class="show-password-icon">
+                  </button>
+                </div>
+                <p class="error-message" data-error-password></p>
+              </div>
+              <hr>
```
Comment (GEOFARL) [inline src/views/pages/sign-in.ejs:49]: style this separators as in design

File: src/client/public/shared/helpers/password-toggle.ts (hunk 13 above)
Comment (GEOFARL) [inline src/client/public/shared/helpers/password-toggle.ts:20]: use dom module that we have in `/client/src/shared/libs/modules/dom/dom.ts`, add a new method to it for changing attributes of DOM elements

**Review by GEOFARL — CHANGES_REQUESTED** [review]

**Comment by kovalsofiia1** [comment]
Знімок екрана 2024-03-28 124915

File: src/client/public/css/components/button.css [hunk 19]
```diff
@@ -9,5 +9,17 @@
 	background-color: var(--color-red);
 	font-size: var(--font-size-700);
 	line-height: 24px;
+	width: 100%;
+	font-weight: 500;
+	cursor: pointer;
+	transition: all 0.6s ease-in-out;
```
Comment (GEOFARL) [inline src/client/public/css/components/button.css:15]:
```suggestion
 transition: all 0.2s ease-in-out;
    border: 2px solid transparent;
```

File: src/client/public/css/components/input.css [hunk 20]
```diff
@@ -10,11 +10,17 @@
...
+	outline: none;
 	font-family: inherit;
 	font-size: inherit;
 	width: 100%;
 	background-color: rgba(175 175 175 / 20%);
 }
+
+.input:hover,
+.input:active {
+	border-color: var(--color-grey-300);
```
Comment (GEOFARL) [inline src/client/public/css/components/input.css:25]:
```suggestion
	border-color: var(--color-grey-300);
  background-color: rgba(175 175 175 /15%);
```

let's add here backgroun change as well

File: src/client/public/css/pages/sign-in.css [hunk 21]
```diff
@@ -9,10 +9,81 @@
...
 .error-message {
 	color: var(--color-red);
 	font-size: var(--font-size-200);
 	margin: 5px 0 3px 5px;
 }
+
+.separator {
+	border: none;
+	border-top: 1px solid rgba(156, 156, 156, 0.68);
+}
```
Comment (GEOFARL) [inline src/client/public/css/pages/sign-in.css:89]: will make a separate component out of it as well

File: src/client/src/pages/sign-in/sign-in.ts [hunk 22]
```diff
@@ -4,6 +4,7 @@ import {
 	userSignInValidationSchema,
 } from "@/modules/user/user.js";
 
+import { initPasswordToggle } from "~/shared/helpers/password-toggle";
```
Comment (GEOFARL) [inline src/client/src/pages/sign-in/sign-in.ts:7]:
```suggestion
import { initPasswordToggle } from "~/shared/index.js";
```

create a file `helpers.ts`, reexport that function from there and then reexport it inside of the `shared/index.ts` from `./libs/helpers/helpers.js`

File: src/client/src/shared/helpers/password-toggle.ts [hunk 23]
```diff
@@ -0,0 +1,29 @@
+import { dom } from "../libs/modules/dom/dom";
+
+const initPasswordToggle = () => {
```
Comment (GEOFARL) [inline src/client/src/shared/helpers/password-toggle.ts:3]: file name should match the name of the function
rename it to the init-password-toggle.ts

File: src/client/src/shared/helpers/password-toggle.ts [hunk 24]
```diff
@@ -0,0 +1,29 @@
+import { dom } from "../libs/modules/dom/dom";
```
Comment (GEOFARL) [inline src/client/src/shared/helpers/password-toggle.ts:1]:
```suggestion
import { dom } from "../libs/modules/dom/dom.js";
```

include js extension in all of the imports

File: src/client/src/shared/helpers/password-toggle.ts [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+
+const initPasswordToggle = () => {
+	const passwordInput = dom.getElement("#password");
+	const togglePasswordBtn = dom.getElement("#togglePassword");
+
+	togglePasswordBtn.addEventListener("click", () => {
+		const type =
+			passwordInput.getAttribute("type") === "password" ? "text" : "password";
+		dom.changeAttribute("#password", "type", type);
```
Comment (GEOFARL) [inline src/client/src/shared/helpers/password-toggle.ts:10]:
```suggestion
```

File: src/client/src/shared/helpers/password-toggle.ts [hunk 26]
```diff
@@ -0,0 +1,29 @@
+import { dom } from "../libs/modules/dom/dom";
+
+const initPasswordToggle = () => {
+	const passwordInput = dom.getElement("#password");
+	const togglePasswordBtn = dom.getElement("#togglePassword");
```
Comment (GEOFARL) [inline src/client/src/shared/helpers/password-toggle.ts:5]:
```suggestion
```

File: src/client/src/shared/helpers/password-toggle.ts [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+
+	togglePasswordBtn.addEventListener("click", () => {
+		const type =
+			passwordInput.getAttribute("type") === "password" ? "text" : "password";
```
Comment (GEOFARL) [inline src/client/src/shared/helpers/password-toggle.ts:9]:
```suggestion
		const type =
			dom.getAttribute({
				attribute: "type",
				selector: "#password",
			}) === "password"
				? "text"
				: "password";
```

File: src/client/src/shared/helpers/password-toggle.ts [hunk 28]
```diff
@@ -0,0 +1,29 @@
+import { dom } from "../libs/modules/dom/dom";
+
+const initPasswordToggle = () => {
+	const passwordInput = dom.getElement("#password");
+	const togglePasswordBtn = dom.getElement("#togglePassword");
+
+	togglePasswordBtn.addEventListener("click", () => {
```
Comment (GEOFARL) [inline src/client/src/shared/helpers/password-toggle.ts:7]:
```suggestion
	 dom.setListener({
    selector: '#toggle-password',
    eventType: 'click',
    listener: () => {}
  })
```

File: src/client/src/shared/helpers/password-toggle.ts [hunk 29]
```diff
@@ -0,0 +1,29 @@
...
+			passwordInput.getAttribute("type") === "password" ? "text" : "password";
+		dom.changeAttribute("#password", "type", type);
+
+		// Change the icon based on the password visibility
+		if (type === "password") {
+			dom.changeAttribute(
+				".show-password-icon",
+				"src",
+				"/icons/eye-closed.svg",
+			);
```
Comment (GEOFARL) [inline src/client/src/shared/helpers/password-toggle.ts:18]:
```suggestion
			dom.changeAttribute({
				selector: ".show-password-icon",
				type: "src",
				value: "/icons/eye-closed.svg",
			});
```

File: src/client/src/shared/libs/modules/dom/base-dom.module.ts [hunk 30]
```diff
@@ -1,6 +1,13 @@
 import { type DOM } from "./libs/types/dom.type.js";
 
 class BaseDOM implements DOM {
+	changeAttribute(selector: string, attribute: string, value: string): void {
```
Comment (GEOFARL) [inline src/client/src/shared/libs/modules/dom/base-dom.module.ts:4]: should be similar to setListener method

File: src/client/src/shared/libs/modules/dom/base-dom.module.ts [hunk 31]
```diff
@@ -1,6 +1,13 @@
...
+		const element = document.querySelector(selector);
+		if (element) {
+			element.setAttribute(attribute, value);
+		}
```
Comment (GEOFARL) [inline src/client/src/shared/libs/modules/dom/base-dom.module.ts:8]: reuse existing methods of the class (getElement)

File: src/views/components/link.ejs [hunk 32]
```diff
@@ -0,0 +1,2 @@
+<a href="<%= href %>"<% if (typeof className !== 'undefined') { %> class="<%= className %>"<% } %>><%= content %></a>
```
Comment (GEOFARL) [inline src/views/components/link.ejs:1]:
```suggestion
<a href="<%- href %>" class="link <%- locals.className ? className : '' %>"><%- content %></a>
```
add default link styles inside of the link.css file

File: src/views/pages/sign-in.ejs [hunk 33]
```diff
@@ -6,37 +6,61 @@
...
+          <div class="prompt">
+            <p class="prompt__question">Немає акаунту?</p>
+            <%- include('../components/link.ejs', {
+              href: "/sign-up",
+              className: "prompt__link",
+              content: "Зареєструватися"
+            }); %>
+          </div>
+        </div>
+        <hr class="separator">
```
Comment (GEOFARL) [inline src/views/pages/sign-in.ejs:20]: make a component out of it

File: src/views/pages/sign-in.ejs [hunk 34]
```diff
@@ -6,37 +6,61 @@
...
+                    <%- include('../components/input.ejs', {
+                      id: "password",
+                      label: "Password",
+                      type: "password",
+                      placeholder: "••••••••",
+                      dataAttributes: {
+                        "data-input-password": ""
+                      }
+                    }); %>
+                    <button class="toggle-password-btn" id="togglePassword" type="button">
```
Comment (GEOFARL) [inline src/views/pages/sign-in.ejs:46]:
```suggestion
                    <button class="toggle-password-btn" id="toggle-password" type="button">
```

**Review by GEOFARL — CHANGES_REQUESTED** [review]

**Comment by GEOFARL** [comment]
@kovalsofiia1, do not include 'fix' when creating a fix commit in the message after :

**Review by GEOFARL — APPROVED** [review]

**Pull request merged**